crates/win-api-wrappers/src/security/acl.rs

```diff
@@ -1,464 +1,265 @@
-
-pub enum AceType {
-    AccessAllowed(Sid),
+//! Helpers to work with ACL structures.
+//!
+//! # Implementation
+//!
+//! Relevant links:
+//! - <https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-acl>
+//! - <https://learn.microsoft.com/en-us/windows/win32/secauthz/creating-or-modifying-an-acl>
+//! - <https://learn.microsoft.com/en-us/windows/win32/secauthz/modifying-the-acls-of-an-object-in-c-->
+//! - <https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-initializeacl>
+
+use core::ptr;
+
+use windows::Win32::Foundation::{LocalFree, HLOCAL};
+use windows::Win32::Security;
+use windows::Win32::System::Memory;
+
+use crate::identity::sid::Sid;
+use crate::utils::u32size_of;
+
+/// Owned ACL freed by LocalFree on drop.
+pub struct Acl {
+    // INVARIANT: Valid pointer to a an initialized ACL structure.
+    // INVARIANT: The pointer must be freed using LocalFree.
+    ptr: HLOCAL,
 }
 
-impl AceType {
-    pub fn kind(&self) -> u8 {
-        // Values for ACE types actually always fit in a u8 even though the type in windows crate is u32.
-        #[expect(clippy::cast_possible_truncation)]
-        match self {
-            AceType::AccessAllowed(_) => ACCESS_ALLOWED_ACE_TYPE as u8,
-        }
-    }
-
-    pub fn to_raw(&self) -> Result<Vec<u8>> {
-        Ok(match self {
-            AceType::AccessAllowed(sid) => RawSid::try_from(sid)?.buf,
-        })
-    }
-
-    pub fn from_raw(kind: u8, buf: &[u8]) -> Result<Self> {
-        let raw_sid = PSID(buf.as_ptr().cast_mut().cast());
+impl Acl {
+    pub fn new() -> windows::core::Result<Self> {
+        // https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-initializeacl
+        // > To calculate the initial size of an ACL, add the following together, and then align the result to the nearest DWORD:
+        // >     Size of the ACL structure.
+        // >     Size of each ACE structure that the ACL is to contain minus the SidStart member (DWORD) of the ACE.
+        // >     Length of the SID that each ACE is to contain.
+        //
+        // The pointer must be aligned with a DWORD = u32.
+        //
+        // The Windows heap managers have always performed heap allocations with a start address that is 8-byte aligned.
+        // (On 64-bit platforms the alignment is 16-bytes).
+        // Microsoft never documented it as a formal guarantee, but given its multi-decade consistency in practice,
+        // it’s essentially become a stable implementation detail.
+
+        // SAFETY: FFI call with no outstanding precondition.
+        let ptr = unsafe { Memory::LocalAlloc(Memory::LMEM_ZEROINIT, size_of::<Security::ACL>())? };
+
+        // SAFETY: The buffer is u32-aligned (= DWORD-aligned), since both 8-byte and 16-byte alignments are stricter.
+        unsafe { Security::InitializeAcl(ptr.0.cast(), u32size_of::<Security::ACL>(), Security::ACL_REVISION)? };
```

Comment:
Does that sound good to you? @thenextman

Reply:
It matches my understanding, from the documentation 👍